gobblin-restli/gobblin-throttling-service/gobblin-throttling-service-client/src/main/java/org/apache/gobblin/util/limiter/BatchedPermitsRequester.java

```diff
@@ -196,18 +205,26 @@ private long elapsedMillis(long startTimeNanos) {
   /**
    * Send a new permit request to the server.
    */
-  private void maybeSendNewPermitRequest() {
-    if (!this.requestSemaphore.tryAcquire()) {
-      return;
+  private synchronized void maybeSendNewPermitRequest() {
+    while (!this.requestSemaphore.tryAcquire()) {
+      if (this.currentCallback == null) {
+        throw new IllegalStateException("Semaphore is unavailable while callback is null!");
+      }
+      if (this.currentCallback.elapsedTime() > 30000) {
```

Comment:
This number seems to be derived from maxTimeOut / (maxRetry + 1) ? Shall we explicitly place a formula here instead of a plain number

Reply:
It is not actually derived from that, it's an arbitrary number. Added some comments and logs though.